autopilot/contractor/contractor.go

```diff
@@ -745,212 +625,178 @@ func renterFundsToExpectedStorage(renterFunds types.Currency, duration uint64, p
+		// create separate usability logger
+		usability := logger.Named("usability").
+			With("contractID", c.ID).
+			With("usability", c.Usability).
+			With("hostKey", c.HostKey)
 
 		// check if contract is ready to be archived.
-		if reason := cc.shouldArchive(c, bh, network); reason != nil {
-			if err := bus.ArchiveContracts(ctx, map[types.FileContractID]string{
-				c.ID: reason.Error(),
-			}); err != nil {
+		if reason := cc.shouldArchive(c, cs.BlockHeight, network); reason != nil {
+			if err := bus.ArchiveContracts(ctx, map[types.FileContractID]string{c.ID: reason.Error()}); err != nil {
 				logger.With(zap.Error(err)).Error("failed to archive contract")
 			} else {
 				logger.Debug("successfully archived contract")
+				usability.Infof("updated contract usability from %s -> bad, reason: %v", c.Usability, reason)
 			}
-			churnReasons[c.ID] = reason.Error()
 			continue
 		}
 
 		// fetch host
 		host, err := bus.Host(ctx, c.HostKey)
 		if err != nil {
 			logger.With(zap.Error(err)).Warn("missing host")
-			churnReasons[c.ID] = api.ErrUsabilityHostNotFound.Error()
+			updateUsability(ctx, cm, api.ContractUsabilityBad, api.ErrUsabilityHostNotFound.Error(), usability)
 			continue
 		}
 
 		// extend logger
-		logger = logger.With("addresses", host.ResolvedAddresses).
+		logger = logger.
+			With("addresses", host.ResolvedAddresses).
 			With("blocked", host.Blocked)
 
 		// check if host is blocked
 		if host.Blocked {
 			logger.Info("host is blocked")
-			churnReasons[c.ID] = api.ErrUsabilityHostBlocked.Error()
+			updateUsability(ctx, cm, api.ContractUsabilityBad, api.ErrUsabilityHostBlocked.Error(), usability)
 			continue
 		}
 
 		// check if host has a redundant ip
 		if ctx.ShouldFilterRedundantIPs() && ipFilter.HasRedundantIP(host) {
 			logger.Info("host has redundant IP")
-			churnReasons[c.ID] = api.ErrUsabilityHostRedundantIP.Error()
+			updateUsability(ctx, cm, api.ContractUsabilityBad, api.ErrUsabilityHostRedundantIP.Error(), usability)
 			continue
 		}
 
 		// get check
 		if host.Checks == (api.HostChecks{}) {
 			logger.Warn("missing host check")
-			churnReasons[c.ID] = api.ErrUsabilityHostNotFound.Error()
+			updateUsability(ctx, cm, api.ContractUsabilityBad, api.ErrUsabilityHostCheckNotFound.Error(), usability)
 			continue
 		}
 
-		// NOTE: if we have a contract with a host that is not scanned, we either
-		// added the host and contract manually or reset the host scans. In that case,
-		// we ignore the fact that the host is not scanned for now to avoid churn.
-		if inSet && host.Checks.UsabilityBreakdown.NotCompletingScan {
-			keepContract(c.ContractMetadata, host)
+		// NOTE: if we have a contract with a host that is not scanned, we
+		// either added the host and contract manually or reset the host scans.
+		// In that case, we ignore the fact that the host is not scanned for now
+		// to avoid churn.
+		if c.IsGood() && host.Checks.UsabilityBreakdown.NotCompletingScan {
+			ipFilter.Add(host)
```

Comment:
you could move this into `updateUsability` and call it every time `api.ContractUsabilityGood` is supplied to make it impossible to miss a host in the filter.